server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1140,11 +1141,23 @@ void runCleanup(ActionListener<DeleteResult> listener) {
         // Updating the shard-level metadata and accumulating results
 
         private void writeUpdatedShardMetadataAndComputeDeletes(ActionListener<Void> listener) {
-            try (var listeners = new RefCountingListener(listener)) {
-                for (IndexId indexId : originalRepositoryData.indicesToUpdateAfterRemovingSnapshot(snapshotIds)) {
-                    new IndexSnapshotsDeletion(indexId).run(listeners.acquire());
-                }
-            }
+            // noinspection resource -- closed safely at the end of the iteration
+            final var listeners = new RefCountingListener(listener);
+
+            // Each per-index process takes some nonzero amount of working memory to hold the relevant snapshot IDs and metadata generations
+            // etc. which we can keep under tighter limits and release sooner if we limit the number of concurrently processing indices.
+            // Each one needs at least one snapshot thread at all times, so threadPool.info(SNAPSHOT).getMax() of them at once is enough to
+            // keep the threadpool fully utilized.
+            ThrottledIterator.run(
```

Comment:
This is a neat class, TIL. In `BlobStoreRepository`, we have 3 ways for running jobs on a thread pool. I wonder whethere there is some possiblity to consolidate them:
* `ThreadPool#executor#execute` - eagerly runs jobs on all available threads and eagerly queue remaining jobs
* `ThrottledTaskRunner#enqueueTask` - Uses up to only a specified number of threads in the underlying threadpoo. Manages its own queue to avoid flooding the threadpool's queue.
* `ThrottledIterator#run` - Uses up to only a specified number of threads in the underlying threadpool. Does not manage its own queue but still avoids flooding the threadpool's queue with an iterator. The usage of iterator also makes it more memory efficient than the above classes because jobs are *not* materialized till they are needed. One important difference with this class is that jobs must be added only once with an `Iterator`. It is not possible to add more jobs afterwards.

I am sure there are details that I am not aware yet. But conceptually it seems useful to have one abstraction that does the following:
1. Configurable max number of threads to use
2. Configurable queuing strategy, aggressive on conservative.
3. Able to take new jobs beyond its instantiation time
4. Eagerly run jobs on a single (or configurable) thread similar to `ThrottledTaskRunner`

Its most basic form would essentially behave just like a `ThreadPool#executor` and the advanced forms would be like `ThrottledTaskRunner` or `ThrottledIterator`. If such thing is possible, we can potentially use it in all places of `BlobStoreRepository`. Right now, we are still using plain `ThreadPool#executor` at shard deletion level. If this usage also needs to be migrated, I wonder whether it is helpful if we could have an overarching Threading abstraction that manages all jobs came out from a snapshot deletion. 

Apologies this feels a bit rambling on my side. Feel free to disagree. Thanks!

Reply:
I'm not sure we need to consolidate these things, they're quite different in many important ways. Conceptually:

- `ThrottledTaskRunner` is about fairness/priority in a threadpool queue.
- `ThrottledIterator` is about limiting work-in-progress and not really anything to do with threading.

It's appropriate to use a plain `ThreadPool#executor` for the shard-level operations, there are effectively no other operations for the threadpool to do at this point (except perhaps some older background cleanup work but that can wait). It's important for the shard-level operations to go through as fast as possible because they're holding up the next snapshot. We don't want to interleave them with any background cleanup work from a previous delete which is what a `ThrottledTaskRunner` would do.

Reply:
> It's important for the shard-level operations to go through as fast as possible because they're holding up the next snapshot.

I realised after writing this that of course this change does have the effect of interleaving the shard-level metadata ops with any leftover background cleanup activity a bit more. Eh. My feeling is that this is ok, background cleanup activity is not normally very onerous, and I believe it's more important to limit the WIP in this area.

Reply:
It's true that `ThrottledIterator` only makes sense if the tasks do not all complete immediately on the calling thread, but that might mean that they dispatch a single task to a threadpool executor, or they dispatch several tasks (which in turn possibly dispatch further tasks) or indeed they send some work to another node entirely.

Originally `ThrottledTaskRunner` did not take a `Releasable` that tasks could use to indicate their completion on a different thread, it was only about whether the forked task had completed execution on the calling thread or not. Now we have that I guess we can cover some similar cases with it. I'd still want it to take an `Iterator` which computes the tasks on the fly, rather than having a fully-materialized queue tho.